libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2130,3 +2130,105 @@ string YulUtilFunctions::revertReasonIfDebug(string const& _message)
 {
 	return revertReasonIfDebug(m_revertStrings, _message);
 }
+
+std::string YulUtilFunctions::decodeReturnParametersFunction(
+	RevertStrings _revertStrings,
+	TypePointers const& _returnTypes,
+	bool _dynamicReturnSize,
+	string const& _retVars
+)
+{
+	solAssert(!_retVars.empty(), "This function must only be called when there are return-parameters to be decoded.");
+
+	ABIFunctions abi(m_evmVersion, _revertStrings, m_functionCollector);
+
+	string const abiDecode = abi.tupleDecoder(_returnTypes, true);
+	string const functionName =
+		"decode_return_parameters_" +
+		revertStringsToString(_revertStrings) + "_" +
+		abiDecode;
+
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return util::Whiskers(R"(
+			function <functionName>(dataMpos, returnSize) <?hasRetVars> -> <retVars> </hasRetVars> {
+				<?dynamicReturnSize>
+					// copy dynamic return data out
+					returndatacopy(dataMpos, 0, returndatasize())
+				</dynamicReturnSize>
+
+				// update freeMemoryPointer according to dynamic return size
+				mstore(<freeMemoryPointer>, add(dataMpos, and(add(returnSize, 0x15), not(0x1f))))
+
+				// decode return parameters from external try-call into retVars
+				<retVars> := <abiDecode>(dataMpos, add(dataMpos, returnSize))
+			}
+		)")
+		("retVars", _retVars)
+		("hasRetVars", !_retVars.empty())
+		("functionName", functionName)
+		("dynamicReturnSize", _dynamicReturnSize)
+		("freeMemoryPointer", to_string(CompilerUtils::freeMemoryPointer))
+		("abiDecode", abiDecode)
+		.render();
+	});
+}
+
+std::string YulUtilFunctions::tryDecodeErrorMessageFunction()
+{
+	string const functionName = "try_decode_error_message";
+
+	return m_functionCollector.createFunction(functionName, [&]() {
+		string const errorHash = FixedHash<4>(util::keccak256("Error(string)")).hex();
+		return util::Whiskers(R"(
+			function <functionName>() -> data {
+				data := mload(0x40)
+				mstore(data, 0)
```

Comment:
What is the purpose of this instruction?

Reply:
TBH, this is part of the initial copy'n'paste from `ContractCompiler::handleCatch(...)`. I preserved that part.

When looking at it, I believe the original authors intention was to default-initialize `data`. However, it's always overwritten later in the lower paths. -> I've removed that line.

Reply:
I really don't know what the original author's mind was thinking back then ;)